feat: new documentation (#164)
tier4/CalibrationTools: sensor/tag_based_pnp_calibrator/README.md

# tag_based_pnp_calibrator

A tutorial for this calibrator can be found [here](../docs/tutorials/tag_based_pnp_calibrator.md)

## Purpose

> The package `tag_based_pnp_calibrator` allows extrinsic calibration among the camera sensor and lidar sensor used in autonomous driving and robotics.

Author: ```suggestion
The package `tag_based_pnp_calibrator` allows extrinsic calibration among the camera and lidar sensors used in autonomous driving and robotics.
```

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/32dd61431d3beb3d84ecb085cd246042de8efb09

## Inner-workings / Algorithms

> The `tag_based_pnp_calibrator` utilizes the PnP algorithm to calculate the transformation between the camera and lidar. To run this package, you also need to operate the `apriltag_ros` package and the `lidartag` package to calculate the transformation.

Author: The verb does not operate here. We do not operate a package/node (in the way we operate a machine). We use, require, or execute

Author: `apriltag_ros` and the `lidartag` packages

Reviewer: fix in https://github.com/tier4/CalibrationTools/pull/164/commits/32dd61431d3beb3d84ecb085cd246042de8efb09

> The `tag_based_pnp_calibrator` utilizes the PnP algorithm to calculate the transformation between the camera and lidar. To run this package, you also need to operate the `apriltag_ros` package and the `lidartag` package to calculate the transformation.

Author: This is the first time PnP is is used to it should be explained

> The `apriltag_ros` package detects the AprilTag and outputs the detection results. Conversely, the `lidartag` package detects the LidarTag and outputs its detection results.

Author: ```suggestion
The `apriltag_ros` package detects AprilTag markers from an image and outputs the detection results. Conversely, the `lidartag` package detects LidarTag markers and outputs its detection results.
```

Reviewer: fix in https://github.com/tier4/CalibrationTools/pull/164/commits/32dd61431d3beb3d84ecb085cd246042de8efb09

The `tag_based_pnp_calibrator` utilizes the detections from both `apriltag_ros` package and `lidartag` package, employing a Kalman Filter to track these detections. If the detections converge, the calibrator applies the SQPnP algorithm provided by OpenCV to estimate the transformation between the image points from AprilTag and the object points from LidarTag.

### Diagram

Below, you can see how the algorithm is implemented in the `tag_based_pnp_calibrator` package.

> ![segment](../docs/images/tag_based_pnp_calibrator/tag_based_pnp_calibrator.jpg)

Author: This diagram is not clear at all in what are the relationship between elements. For example, it is not an UML, so you need to explain the kind of connections you have (I get the feeling that different kind of connections are used with the same line type). I will only write it here, but probably applies to all your diagrams

| ----------------------- | ----------------------------------------------------- | ---------------------------------------------------------------------------------------- |
| `extrinsic_calibration` | `tier4_calibration_msgs::` `srv::ExtrinsicCalibrator` | Generic calibration service. The call is blocking until the calibration process finishes |

## Parameters

### Core Parameters

| Name                                          | Type                  | Default Value | Description                                                                                                                         |
| --------------------------------------------- | --------------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------- |
| `calib_rate`                                  | `double`              |               | The rate at which the calibration callback is invoked. This controls the frequency of calibration updates.                          |

> | `base_frame`                                  | `std::string`         |               | The base_frame is used to compare the initial and calibrated values.                                                                |

Author: ```suggestion
| `base_frame`                                  | `std::string`         |               | The base_frame is used for visualization.                                                                |
```

> | `min_tag_size`                                | `double`              |               | The size of the AprilTag in meters                                                                                                  |

Author: ```suggestion
| `min_tag_size`                                | `double`              |               | The size of the AprilTag in meters (payload)                                                                                                  |
```

> | `max_tag_distance`                            | `double`              |               | Maximum allowable distance from the camera to the tags.                                                                             |

Author: In this context it should be allowed

Author: This was addressed in some commit already

> | `max_allowed_homography_error`                | `double`              |               | Discard the AprilTag detection if the homography error is larger than this value.                                                   |

Author: ```suggestion
| `max_allowed_homography_error`                | `double`              |               | AprilTag detection are discarded if the homography error is larger than `max_allowed_homography_error`.                                                   |
```

| `use_receive_time`                            | `bool`                |               | Flag to determine whether to use the receive time instead of the header timestamps.                                                 |
| `use_rectified_image`                         | `bool`                |               | Flag to determine whether rectified images should be used in the calibration process.                                               |

> | `calibration_crossvalidation_training_ratio`  | `double`              |               | The ratio of data used for training versus testing during the calibration's cross-validation process.                               |

Author: It is either validation or testing...

Reviewer: @knzo25  Would like to ask what do you mean for this comment. I wrote testing.

Author: @vividf 
Testing and validation have both different meanings yet depending on where you learned and from whom you learned, they have their meanings changed. Since you are already using cross-validation, you can not use testing

Reviewer: Solved in https://github.com/tier4/CalibrationTools/pull/164/commits/f74fdb89b724f96b096c8f959a82787c77d706e8

| `calibration_convergence_min_pairs`           | `int`                 |               | The minimum number of AprilTag and LidarTag detection pairs required to consider the calibration process as potentially converging. |
| `calibration_convergence_min_area_percentage` | `double`              |               | Minimum percentage of the area that needs to be covered by detection.                                                               |
| `min_pnp_points`                              | `int`                 |               | Minimum number of points required for the Perspective-n-Point problem used in calibration to solve the pose estimation.             |
| `min_convergence_time`                        | `double`              |               | Minimum time required for the calibration process to be considered as converged.                                                    |

> | `max_no_observation_time`                     | `double`              |               | Maximum allowable time interval without observing any calibration tags before considering the calibration data stale.               |

Author: This is wrong

| `new_hypothesis_distance`                     | `double`              |               | Distance threshold for creating a new hypothesis.                                                                                   |
| `tag_ids`                                     | `std::vector<int>`    |               | List of tag IDs that are used in the calibration process.                                                                           |

> | `tag_sizes`                                   | `std::vector<double>` |               | Physical sizes of the tags corresponding to the IDs in meter.                                                                       |

Author: It is not the physical size. It is the payload. They are different in this case

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

> | `lidartag_max_convergence_translation`        | `double`              |               | Maximum translation allowed for a lidar tag detection hypothesis to be considered converging.                                       |

Author: Both "to be considered converging" and "considered as converged" are used. Normalize

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

| `lidartag_max_convergence_translation_dot`    | `double`              |               | Maximum translation dot allowed for a lidar tag detection hypothesis to be considered converging.                                   |
| `lidartag_max_convergence_rotation`           | `double`              |               | Maximum rotation allowed for a lidar tag detection hypothesis to be considered converging.                                          |

> | `lidartag_max_convergence_rotation_dot`       | `double`              |               | Maximum rotation dot allowed for a lidar tag detection hypothesis to be considered converging.                                      |

Author: Do you know what dot means here?

Reviewer: Yes, I forget to change it to velocity!
Thanks!
Fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

| `lidartag_new_hypothesis_translation`         | `double`              |               | Translation threshold for generating a new hypothesis in lidar tag tracking.                                                        |
| `lidartag_new_hypothesis_rotation`            | `double`              |               | Rotation threshold for generating a new hypothesis in lidar tag tracking.                                                           |

> | `lidartag_measurement_noise_translation`      | `double`              |               | The square of this value is part of input for cv::KalmanFilter::measurementNoiseCov to measurement noise covariance matrix (R).     |

Author: You do not need to specify the implementation (cv) of the Kalman Filter. Saying what part of the kalman filter is suffices. The units are missing

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

| `lidartag_process_noise_rotation_dot`         | `double`              |               | The square of this value is part of input for cv::KalmanFilter::processNoiseCov to process noise covariance matrix (Q).             |
| `apriltag_max_convergence_translation`        | `double`              |               | Maximum translation error allowed for an AprilTag detection hypothesis to be considered converging.                                 |
| `apriltag_new_hypothesis_translation`         | `double`              |               | Translation threshold for generating a new hypothesis in AprilTag tracking.                                                         |
| `apriltag_measurement_noise_translation`      | `double`              |               | Estimated noise level for translation measurements in AprilTag tagging.                                                             |
| `apriltag_process_noise_translation`          | `double`              |               | The square of this value is part of input for cv::KalmanFilter::processNoiseCov to process noise covariance matrix (Q).             |

## Requirements

### LiDARTag

> 

Author: I am pretty sure we can put more default values in here

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

> To perform camera-lidar calibration using this tool, it is necessary to prepare LidarTags and lidars with intensity measures. To ensure that no objects obstruct the tag detection and to achieve the most stable detection possible, it is highly recommended to also prepare fixed mounts for these tags, as shown below.

Author: Some paragraph that explains the size recommendations are missing.

The ones we use are lidartags of size 0.8m. Meaning the apriltag payload is 0.6m
We have also tried with lidartag sizes of 0.6m, but to use them the user needs to set several parameters by himself

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

<p align="center">
    <img src="../docs/images/tag_based_pnp_calibrator/lidartag-mount.jpg"  alt="lidartag-mount" width="500">
</p>

## References

References/External links
[1] Jiunn-Kai (Bruce) Huang, Shoutian Wang, Maani Ghaffari, and Jessy W. Grizzle, "LiDARTag: A Real-Time Fiducial Tag System for Point Clouds," in IEEE Robotics and Automation Letters. Volume: 6, Issue: 3, July 2021.

## Known issues/limitations

> - The tool is currently using the pinhole model for calibration. If the user is using the fisheye camera, the calibration result will not be perfect.

Author: ```suggestion
- The tool uses a basic opencv camera model for calibration (plumb bomb).
```

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

## Pro tips/recommendations

> - During calibration, ensure that the lidar scan covers the tag, similar to the first example shown in the image below. However, if the tag resolution is low, as in the second example, and the lidar still detects the tag, it is acceptable. The third example demonstrates a scenario where the lidar scan fails to cover the tag, resulting in the inability to detect the LidarTag.

Author: ```suggestion
- During calibration, ensure that the lidar scan covers the tag, similar to the first example shown in the image below. However, if the tag resolution is low, as in the second example, and the lidar still detects the tag, it may still be acceptable, but avoidable when possible. The third example demonstrates a scenario where the lidar scan fails to cover the tag, resulting in the inability to detect the LidarTag.
```

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

<p align="center">
    <img src="../docs/images/tag_based_pnp_calibrator/lidarscan_on_tag.jpg"  alt="lidarscan_on_tag" width="500">
</p>

> - Also note that when doing the calibration, it is necessary to rotate the tag facing the camera like the image shown below.

Author: ```suggestion
- It is highly recommended to place the tag perpendicular to the calibration lidar as shown in the following image:
```

Reviewer: Thanks! fixed in https://github.com/tier4/CalibrationTools/commit/32dd61431d3beb3d84ecb085cd246042de8efb09

# tag_based_pnp_calibrator

A tutorial for this calibrator can be found [here](../docs/tutorials/tag_based_pnp_calibrator.md).

## Purpose

> The package `tag_based_pnp_calibrator` allows extrinsic calibration among the camera and lidar sensors used in autonomous driving and robotics.

Author: ```suggestion
The package `tag_based_pnp_calibrator` allows extrinsic calibration among camera and 3d lidar sensors used in autonomous driving and robotics.
```

## Inner-workings / Algorithms

> The `tag_based_pnp_calibrator` utilizes the PnP (Perspective-n-Point) algorithm, a computer vision method that finds the best match between a set of 3D points (from the lidar) and their corresponding 2D projections (in the camera images), to accurately calculate the transformation between the camera and lidar. To run this package, you also need to execute the `apriltag_ros` and the `lidartag` packages to calculate the transformation.

Author: PnP does not find the best match. The matches are "input" of the algorithm and are fixed during optimization

Author: The way this paragraph is written, it implies the user must run the extra packages manually. It could be made explicit that the package's launcher also launches those as needed

Author: @vividf 
friendly reminder

Reviewer: Thanks, fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/765904534a790bdf1517564399b3eed8de259f20

Author: I will resolve this for now and write a suggestion

> The `apriltag_ros` package detects AprilTag markers from an image and outputs the detection results. Conversely, the `lidartag` package detects LidarTag markers and outputs its detection results.

Author: I see:
 - LiDARTag
 - LidarTag
 - Lidartags
 - lidartag

Please make it consistent
Although I am inclined to use `lidartag` (due to ros2 naming conventions), I will leave the choice to you.

Author: @vividf 
friendly reminder

Reviewer: @knzo25 
Thanks! After I fixed them, I didn't see this comment.
https://github.com/tier4/CalibrationTools/pull/164/commits/8cc5d621ab26a27f6d02d53d64d754a50b2b3601

> The `tag_based_pnp_calibrator` utilizes the detections from both `apriltag_ros` package and `lidartag` package, employing a Kalman Filter to track these detections. If the detections converge, the calibrator applies the SQPnP algorithm provided by OpenCV to estimate the transformation between the image points from AprilTag and the object points from LidarTag.

Author: (optional)
Reference to the sqpnp paper

Author: For optional comments, if you do not agree, you can either state the reason so that I `resolve` it, or you can `resolve` it yourself

Reviewer: Thanks!
fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/002a39c2ca0808ab85521e054657559c817f4bae

Author: This is not a paper, and we are not following any citation standards really, but there is relevant information missing, like where and when it was published

https://link.springer.com/chapter/10.1007/978-3-030-58452-8_28#citeas

@vividf

Reviewer: added them in https://github.com/tier4/CalibrationTools/pull/164/commits/b0b7c86085e42a8616f632c92b0915cf761cd779

Author: That empty () is a format error

Reviewer: removed in https://github.com/tier4/CalibrationTools/pull/164/commits/3b37a80fa659ae4a00b3c60d2072a000a22f4822

> The `tag_based_pnp_calibrator` utilizes the detections from both `apriltag_ros` package and `lidartag` package, employing a Kalman Filter to track these detections. If the detections converge, the calibrator applies the SQPnP algorithm provided by OpenCV to estimate the transformation between the image points from AprilTag and the object points from LidarTag.

Author: "packages" is repeated
Kalman is a filter, not a tracker (it is used in trackers but not a tracker)

Author: @vividf 
friendly reminder

Reviewer: Thanks! I modify "track" to "monitor", hope this fits better. 
fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/3dd72cfe196038823a6a7f11dc6ae63f8e28772a

### Services

| Name                    | Type                                                  | Description                                                                               |
| ----------------------- | ----------------------------------------------------- | ----------------------------------------------------------------------------------------- |
| `extrinsic_calibration` | `tier4_calibration_msgs::` `srv::ExtrinsicCalibrator` | Generic calibration service. The call is blocking until the calibration process finishes. |

## Parameters

### Core Parameters

| Name                                          | Type                  | Default Value                    | Description                                                                                                                                               |
| --------------------------------------------- | --------------------- | -------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `calib_rate`                                  | `double`              | `10.0`                           | The rate at which the calibration callback is invoked. This controls the frequency of calibration updates.                                                |

Author: The frequency units are missing

Author: @vividf 
friendly reminder

Reviewer: Thanks!
fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/aeea36d184fedf3d789d280dfd1d78c482df7f33

Author: Frequency is in hertz !

Reviewer: ah sorry!
fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/b0b453dab4c6dcd853f1b6a6584f41aa084517b5

> | `base_frame`                                  | `std::string`         | `base_link`                      | The base_frame is used for visualization.                                                                                                                 |

Author: `base_frame`

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/e8385df82fd7ae6085d400ffebc8bac44cfa2dd6

| `min_tag_size`                                | `double`              | `0.6`                            | The size of the AprilTag in meters (payload).                                                                                                             |
| `max_tag_distance`                            | `double`              | `20.0`                           | Maximum allowed distance in meters from the camera to the tags.                                                                                           |
| `max_allowed_homography_error`                | `double`              | `0.5`                            | AprilTag detection are discarded if the homography error is larger than `max_allowed_homography_error`.                                                   |
| `use_receive_time`                            | `bool`                | `false`                          | Flag to determine whether to use the receive time instead of the header timestamps.                                                                       |

> | `use_rectified_image`                         | `bool`                | `true`                           | Flag to determine whether rectified images should be used in the calibration process.                                                                     |

Author: The explanation for `use_rectified_image` is wrong.
Just changing the value of this parameter does not change the image as the description implies. It determines if the image is treated as rectified or not

Reviewer: fixed in https://github.com/tier4/CalibrationTools/commit/50eca415a164ec242cdb629eaacf73654c155162

| `calibration_crossvalidation_training_ratio`  | `double`              | `0.7`                            | The ratio of data used for training versus testing during the calibration's cross-validation process.                                                     |
| `calibration_convergence_min_pairs`           | `int`                 | `9`                              | The minimum number of AprilTag and LidarTag detection pairs required to consider the calibration process as potentially converged.                        |
| `calibration_convergence_min_area_percentage` | `double`              | `0.005`                          | Minimum percentage of the area that needs to be covered by detection.                                                                                     |
| `min_pnp_points`                              | `int`                 | `8`                              | Minimum number of points required for the Perspective-n-Point problem used in calibration to solve the pose estimation.                                   |

> | `min_convergence_time`                        | `double`              | `6.0`                            | Minimum time required for the calibration process to be considered as converged.                                                                          |

Author: The description for `min_convergence_time` is wrong

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/35362526d946236ad4f54d91cf0fda99f20bcf57

> | `max_no_observation_time`                     | `double`              | `3.0`                            | If the time between new detection and the last observed detection are larger than `max_no_observation_time`, remove the detection from active hypotheses. |

Author: The description for `max_no_observation_time` is wrong

Reviewer: Sorry, could you tell me the correct explanation? Thanks

Author: Hypotheses can time out in two ways:
 - When a new observation comes in and the dt is higher than the threshold (same tag id)
 - When the latest observation and a particular observation have a dt higher than the threhold.
 
 In the past I used rclcpp::Time() to use the latest time, but there were errors in some use cases (at the time we still  had not decided on our policies regarding sim_time, so things broke a lot. Hence, I decided to use the latest detection stamp)

Reviewer: fixed it in https://github.com/tier4/CalibrationTools/pull/164/commits/3b37a80fa659ae4a00b3c60d2072a000a22f4822.

> | `new_hypothesis_distance`                     | `double`              | `1.5`                            | Distance threshold for creating a new hypothesis.                                                                                                         |

Author: The description for `new_hypothesis_distance` is insufficient. Distance with respect to what

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/c9cfdac34e81a123684be1bfd4899ccd61444437

| `tag_ids`                                     | `std::vector<int>`    | `[0, 1, 2, 3, 4, 5]`             | List of tag IDs that are used in the calibration process.                                                                                                 |

> | `tag_sizes`                                   | `std::vector<double>` | `[0.6, 0.6, 0.6, 0.6, 0.6, 0.6]` | Payload sizes of the tags corresponding to the IDs in meters.                                                                                             |

Author: May be worth to specify that it needs to be of the same size that of the previous parameter (tag_ids)

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/97e2c131f7d3451b61791f4464f5daa56a1c3ad8

> | `lidartag_max_convergence_translation`        | `double`              | `0.05`                           | Maximum translation in meters allowed for a lidar tag detection hypothesis to be considered as converged.                                                 |

Author: The "convergence" parameters are not well explained.
How do you use the parameter? With exactly what you compare it to determine convergence?

Reviewer: I kind of want to write like this "`lidartag_max_convergence_translation` serves as a threshold to define the maximum allowable standard deviation of position estimates (in meters) for the hypothesis being tracked. If the estimated position's uncertainty (standard deviation) is below this threshold, the hypothesis is considered stable enough to be "converged." but I am still not sure if it is the description you want. 

Also, regarding to the code in lidartag_hypothesis.cpp line 205.
The max_translation_dot_cov is always 0.0, I am not sure if this is correct or not.

`  if (
    std::sqrt(max_translation_cov) > convergence_translation_ ||
    std::sqrt(max_translation_dot_cov) > convergence_translation_dot_ ||
    std::sqrt(max_rotation_cov) > convergence_rotation_ ||
    std::sqrt(max_translation_dot_cov) > convergence_rotation_dot_ ||
    getSpeed() > convergence_translation_dot_) {
    return false;
  }`

If possible, could you write these parameters descriptions as I think it's a bit hard for me to explain them perfectly. 
Thanks!

Author: Addressed in [ca0f677](https://github.com/tier4/CalibrationTools/pull/164/commits/ca0f677e403501ca7f66c47f0ba0c0eb42bfbdf1)

| `apriltag_process_noise_translation`          | `double`              |               | The square of this value is part of input for cv::KalmanFilter::processNoiseCov to process noise covariance matrix (Q).             |

## Requirements

### LiDARTag

To perform camera-lidar calibration using this tool, it is necessary to prepare LidarTags and lidars with intensity measures. To ensure that no objects obstruct the tag detection and to achieve the most stable detection possible, it is highly recommended to also prepare fixed mounts for these tags, as shown below.

<p align="center">
    <img src="../docs/images/tag_based_pnp_calibrator/lidartag-mount.jpg"  alt="lidartag-mount" width="500">
</p>

## References

References/External links
[1] Jiunn-Kai (Bruce) Huang, Shoutian Wang, Maani Ghaffari, and Jessy W. Grizzle, "LiDARTag: A Real-Time Fiducial Tag System for Point Clouds," in IEEE Robotics and Automation Letters. Volume: 6, Issue: 3, July 2021.

## Known issues/limitations

Our version of LidarTag only supports the family `16h5`

> Our codebase only supports AprilTag detections for `36h11`

Author: This is not true

Reviewer: removed in commit https://github.com/tier4/CalibrationTools/pull/164/commits/992f44cdbd9a1e720f31f88c707c4556468f96c6

## Pro tips/recommendations

> - During calibration, ensure that the lidar scan covers the tag, similar to the first example shown in the image below. However, if the tag resolution is low, as in the second example, and the lidar still detects the tag, it may still be acceptable, but avoidable when possible. The third example demonstrates a scenario where the lidar scan fails to cover the tag, resulting in the inability to detect the LidarTag.

Author: avoidable when possible?

> - During calibration, ensure that the lidar scan covers the tag, similar to the first example shown in the image below. However, if the tag resolution is low, as in the second example, and the lidar still detects the tag, it may still be acceptable, but avoidable when possible. The third example demonstrates a scenario where the lidar scan fails to cover the tag, resulting in the inability to detect the LidarTag.

Author: ```suggestion
- During calibration, ensure that the lidar scan covers the tag, similar to the first example shown in the image below. However, if the tag resolution is low, as in the second example, and the lidar still detects the tag, it may still be acceptable, but should be avoided when possible. The third example demonstrates a scenario where the lidar scan fails to cover the tag, resulting in the inability to detect the LidarTag.
```

Reviewer: Thanks! fixed in (https://github.com/tier4/CalibrationTools/pull/164/commits/2dee0c576e69b31970d190961530dc63acb91779)

# tag_based_pnp_calibrator

A tutorial for this calibrator can be found [here](../docs/tutorials/tag_based_pnp_calibrator.md).

## Purpose

The package `tag_based_pnp_calibrator` allows extrinsic calibration among camera and 3d lidar sensors used in autonomous driving and robotics.

## Inner-workings / Algorithms

> The `tag_based_pnp_calibrator` utilizes the PnP (Perspective-n-Point) algorithm, a computer vision technique that be used to estimate the transformation between a set of correspondences of 3d and 2d points (in this case the 3d points come from the lidar and the 2d points come from the camera). For the calibration process, both `apriltag_ros` and `lidartag` packages are required, and they are executed automatically by this package's launcher files.

Author: "that be used"

Reviewer: @knzo25 What do you mean?

Author: "...  a computer vision technique that be used to estimate the transformation ..."

Author: @vividf you did not answer this one

Reviewer: Sorry, I didn't make this commit separately before, 
I fixed it in https://github.com/tier4/CalibrationTools/pull/164/commits/f49454cead9f1321aae2e09e9901cdcbe6175489. 

 "that be used" -> "that is used"

Author: @vividf 
Remember to split the commits...

> The `tag_based_pnp_calibrator` utilizes the PnP (Perspective-n-Point) algorithm, a computer vision technique that calculates the transformation between camera and lidar by solving for the pose of a set of 3D points (from the lidar) relative to their corresponding 2D projections (in the camera images). To operate this package, both `apriltag_ros` and `lidartag` packages are required. The package's launcher will automatically initiate these dependencies.

Author: ```suggestion
The `tag_based_pnp_calibrator` utilizes the PnP (Perspective-n-Point) algorithm, a computer vision technique that be used to estimate the transformation between a set of correspondences of 3d and 2d points (in this case the 3d points come from the lidar and the 2d points come from the camera). For the calibration process, both `apriltag_ros` and `lidartag` packages are required, and they are executed automatically by this package's launcher files.
```

The `apriltag_ros` package detects apriltag markers from an image and outputs the detection results. Conversely, the `lidartag` package detects lidartag markers and outputs its detection results.

> The `tag_based_pnp_calibrator` utilizes the detections from both `apriltag_ros` and `lidartag` packages, employing a Kalman Filter to monitor these detections. If the detections converge, the calibrator applies the SQPnP [2] algorithm provided by OpenCV to estimate the transformation between the image points from apriltag and the object points from lidartag.

Author: Kalman filter

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/97adf8745f23ea64ffe7aa830441cd7800c7da7f

### Diagram

Below, you can see how the algorithm is implemented in the `tag_based_pnp_calibrator` package.

![segment](../docs/images/tag_based_pnp_calibrator/tag_based_pnp_calibrator.jpg)

## ROS Interfaces

### Input

| Name                        | Type                                         | Description                                                                                  |
| --------------------------- | -------------------------------------------- | -------------------------------------------------------------------------------------------- |

> | `{camera_info}`             | `sensor_msgs::msg::CameraInfo`               | Intrinsic parameters for the calibration cameras . `camera_info` is provided via parameters. |

Author: You copied this format from the SfM. In that case, support for several cameras was implemented. In this case, it is incorrect

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/dc5164fd5f105480f50f32e7ad4628871da0b194

Author: Parameter is a countable noun.
In addition, is is not a parameter.

Please check if the things you are writing are write before pushing. Are you using proper spell/grammar tools?

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/08a377f424fbd9f56a3e825592ab865fc6bbff3f

Reviewer: I was using the tool before I wrote in parameter because the user can define them in the launcher.

Author: @vividf 
It is still wrong (in a different way this time)

Reviewer: @knzo25 
| `camera_info`               | `sensor_msgs::msg::CameraInfo`               | Intrinsic parameters for the calibration camera. `camera_info` is defined in the launcher. |

Could you tell me which part is wrong? Thanks

Author: `Calibration camera` is a concept of the sfm where we had the external camera and the calibration cameras.
Here there is only one camera; the camera to be calibrated, so it is redundant.
Unlike the sfm, in which all the topics were passed via parameters, here the topic names are defined in the code. As it is written it is weird. If you see the launchers, there is remapping, not definition

Reviewer: I fixed the first part here: https://github.com/tier4/CalibrationTools/pull/164/commits/79c9e471a8112ea5f39ea991ffdec2fc8a1ad96e

Reviewer: For the description part, should I say it is remapping from the parameter "camera_info_topic"?

Author: When topics are created like `this->create_subscriber("topic_name",...)`, we only need to explain the contents of `topic_name`, not explain where it is defined. In particular, it is not defined in the launchers (at most, it is remaped to variable that is defined in the launcher).

The reason that it was needed in the sfm, is because the number of topics was not fixes, and as such, the list of topics was retrieved from a parameter, and thus further explanations were needed. Removing the redundant (and half incorrect part) should be applied in all the places where it is applicable.

In addition:
`Intrinsic parameters for the camera` -> `Intrinsic parameters of the camera`

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/f2e9bb37383fe24008b84ca421b1fed1963167d7 and https://github.com/tier4/CalibrationTools/pull/164/commits/793a581a0174ea4512d7dabed106cae1e1442271

Author: @vividf 
Like the Sfm, the number of calibration lidars is variable.
`calibration_pointcloud_topics` is not a topic, is a parameter. Try to understand the context of comments and code, and then apply where it corresponds. It is not a mechanical process

Reviewer: Thanks, fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/ffcf7088589dd8ec39c75e48ad2eafcb822896ad. I changed the docker by accident so I fixed it.

Author: `Recommend to select the lidar that has the highest resolution.` is not grammatically correct

Reviewer: fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/8d3e0512eaa6efdfedf5dc400b8f4aaf63656f89

Author: `processes` should be deleted.
On another note, the criteria is not only the highest resolution but also the best FOV.
For example, you can not map with a lidar in a low position, nor one that a narrow FOV

Reviewer: thanks, fixed in https://github.com/tier4/CalibrationTools/pull/164/commits/f63c2276aee893ba7b8fdaef978444505d8743cb

> | `lidartag/detections_array` | `lidartag_msgs::msg::LidarTagDetectionArray` | Lidartag detections. `lidartag/detections_array` is defined in launcher.                     |

Author: Did you not unify the spelling for lidartag?

> | `apriltag/detection_array`  | `apriltag_msgs::msg::AprilTagDetectionArray` | AprilTag detections. `apriltag/detection_array` is defined in launcher.                      |

Author: Did you not unify the spelling for apriltag?
